components/popover/positionCalculation.js

```diff
@@ -92,37 +79,75 @@ const updateHorizontalPositionIfNeeded = (position, anchorPosition, targetPositi
-  left: anchorPosition.left - offsetCorrection,
-  arrowLeft: 2 * POPUP_OFFSET,
-});
-
-const positionRight = (anchorPosition, targetPosition, offsetCorrection) => ({
-  left: anchorPosition.right - targetPosition.width + offsetCorrection,
-  arrowLeft: targetPosition.width - POPUP_OFFSET * 3,
-});
+const getPositionTopValue = ({ direction, anchorPosition, targetPosition }) =>
+  direction === 'north'
+    ? anchorPosition.Top - targetPosition.height - POPUP_OFFSET
+    : anchorPosition.top + targetPosition.height + POPUP_OFFSET;
+
+const getPositionLeftValue = ({ renderPosition, anchorPosition, targetPosition, inputOffsetCorrection }) => {
+  switch (renderPosition) {
+    case 'center':
+      return anchorPosition.center - targetPosition.width / 2;
+    case 'left':
+      return anchorPosition.left - inputOffsetCorrection;
+    case 'right':
+      return anchorPosition.right - targetPosition.width + inputOffsetCorrection;
+  }
+};
 
-const directionNorth = (anchorPosition, targetPosition) => ({
-  top: anchorPosition.top - targetPosition.height - POPUP_OFFSET,
-  arrowTop: targetPosition.height - ARROW_OFFSET,
-});
+const getArrowPositionTopValue = ({ direction, targetPosition }) =>
+  direction === 'north' ? targetPosition.height - ARROW_OFFSET : -ARROW_OFFSET;
+
+const getArrowPositionLeftValue = ({ renderPosition, targetPosition }) => {
+  switch (renderPosition) {
+    case 'center':
+      return targetPosition.width / 2 - ARROW_OFFSET;
+    case 'left':
+      return 2 * POPUP_OFFSET;
+    case 'right':
+      return targetPosition.width - POPUP_OFFSET * 3;
+  }
+};
 
-const directionSouth = anchorPosition => ({
-  top: anchorPosition.top + anchorPosition.height + POPUP_OFFSET,
-  arrowTop: -ARROW_OFFSET,
-});
+const getDirectionRendersOnScreen = ({ direction, anchorPosition, targetPosition }) => {
+  switch (direction) {
+    case 'north':
+      return anchorPosition.top - targetPosition.height - POPUP_OFFSET > 0;
+    case 'south':
+      return anchorPosition.bottom + targetPosition.height + POPUP_OFFSET < window.innerHeight;
+    case 'east':
+      return anchorPosition.right + targetPosition.width + POPUP_OFFSET < window.innerWidth;
+    case 'west':
+      return anchorPosition.left - targetPosition.width - POPUP_OFFSET > 0;
+  }
+};
 
-const updateDirectionIfNeeded = (direction, anchorPosition, targetPosition) => {
-  if (direction === 'north') {
-    return anchorPosition.top - targetPosition.height - POPUP_OFFSET < 0 ? 'south' : 'north';
+const getOppositeDirection = direction => {
+  switch (direction) {
+    case 'north':
+      return 'south';
+    case 'south':
+      return 'north';
+    case 'east':
+      return 'west';
+    case 'west':
+      return 'east';
   }
+};
 
-  return anchorPosition.bottom + targetPosition.height + POPUP_OFFSET > window.innerHeight ? 'north' : 'south';
+const getOppositeDirectionRendersOnScreen = ({ direction, anchorPosition, targetPosition }) =>
```

Comment:
Maybe you can eliminate this function and use `isInViewport` with a different `direction` parameter?